### PR TITLE
Use the cache strategy to merge fragment caching into the main page

### PR DESCRIPTION
**Pull request by aschempp**

### CODE REVIEW

**Comment by leofeyer** [comment]
LGTM, although I do not 100% understand how the code works. 🙈

**Review by Toflar — DISMISSED** [review]
LGTM. Apart from missing tests ;)

**Comment by leofeyer** [comment]
So what exactly does this do?

**Comment by aschempp** [comment]
In a default Symfony setup with a Symfony cache, there are „inline“ subrequests and „esi“ subrequests. The point of using ESI is that a fragment can have different caching settings than the main page content. 

However, using ESI for that is only really useful if the fragment differs per user (e.g. a shopping cart), or if generating the fragment is expensive and it can be cached for a long duration. 

In our case, we have a lot of fragments that are neither of both. A `{{date:Y}}` is the same for all users, and it is very inexpensive. It however does affect the page cache time, as the inserttag _could_ be cached much longer than the page usually itself. But instead of using ESI, it is much more efficient if the content **and** cache duration of the main page and fragment are merged.

Example 1:
- today is 2019-12-30 17:00:00
- the page can be cached for 2hrs
- the date insert tag can be cached until 2019-12-31 23:59:59
- **merging both, the content can still be cached for 2hrs, without the need to parse an ESI fragment on every request**

Example 2:
- today is 2019-12-31 23:00
- the page can be cached for 2hrs
- the date insert tag can be cached until 2019-12-31 23:59:59
- **with this PR, the page cache time will be reduced to 2019-12-31 23:59:59, and a second later a new cache response will be generated with the new year, and cached again for 2hrs**

The PR simply merges caching time of inline fragments using the Symfony Cache Strategy (which was optimized by me recently) to apply the same as the Symfony cache does when merging multiple ESI fragments.

File: core-bundle/src/Fragment/FragmentHandler.php
Comment (leofeyer) [inline core-bundle/src/Fragment/FragmentHandler.php]: I think this can be reduced to just `$this->currentStrategy = array_pop($this->strategyStack);` (see https://3v4l.org/fGQGb).

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/Fragment/FragmentHandler.php
Comment (aschempp) [reply core-bundle/src/Fragment/FragmentHandler.php]: Changed in 6dac1ef004af2f2ecfa50f7c0d383d5d4adf8119

**Comment by aschempp** [comment]
Maybe we should not that this first needs actual manual testing before merge to see if what we plan here makes sense and works for the insert tags fragments 🙈

**Comment by leofeyer** [comment]
Ok, then I will not merge it until you have confirmed.

**Comment by aschempp** [comment]
I have updated the PR per discussion with @Toflar 

Instead of letting the `FragmentHandler` handle the requests, we can simply handle them with a kernel event listener. This means **all** requests can add caching informations, even without using the Contao fragment handler.

However, "unfortunately", a Symfony response defaults to `Cache-Control: private`. Therefore, we can't simply rely on the cache-control header to find out if the response should be merged. One way to approach this is a custom response header. To simplify doing that, I have update the `Template::getResponse()` class. It automatically enables caching but removes any existing configuration. In a custom fragment, you can do something like
```
return $template->getResponse()->setMaxAge(100);
```

This PR also improves the `Template::getResponse()` method. It previously always added the `global $objPage` caching infos, which means the method was unusable for anything than the page response. The caching info is now only added if the template name starts with `fe_` or the explicit argument is set.

File: core-bundle/src/EventListener/SubrequestCacheListener.php
Comment (Toflar) [inline core-bundle/src/EventListener/SubrequestCacheListener.php]: Imho this should be `Contao-Merge-Cache-Control`.

File: core-bundle/src/EventListener/SubrequestCacheListener.php
Comment (Toflar) [inline core-bundle/src/EventListener/SubrequestCacheListener.php]: It would be helpful to have a comment here so a new developer understands what this listener is doing :)

File: core-bundle/src/EventListener/SubrequestCacheListener.php
Comment (Toflar) [inline core-bundle/src/EventListener/SubrequestCacheListener.php:46]: Why do we need a stack here? As far as I understood from our latest discussions there's only one master request in the same process unless you're using `HttpCache` in which case the services are reset for every master request. So imho, it should implement the `ResetInterface`.

File: core-bundle/src/Resources/contao/classes/FrontendTemplate.php
Comment (Toflar) [inline core-bundle/src/Resources/contao/classes/FrontendTemplate.php:82]: What's this new parameter for?

File: core-bundle/src/Resources/contao/classes/FrontendTemplate.php
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/FrontendTemplate.php]:
```suggestion
		if ($blnForceCacheHeaders || 0 === strncmp('fe_', $this->strTemplate, 3)) {
```

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/classes/FrontendTemplate.php
Comment (leofeyer) [reply core-bundle/src/Resources/contao/classes/FrontendTemplate.php:82]: I guess @aschempp forgot to add it to the if condition (see [my suggestion](https://github.com/contao/contao/pull/718/files#r353684908)).

File: core-bundle/src/EventListener/SubrequestCacheListener.php
Comment (leofeyer) [inline core-bundle/src/EventListener/SubrequestCacheListener.php]: We could store the result in an `$isMasterRequest` variable and reuse it on line 59. WDYT?

File: core-bundle/src/Resources/contao/classes/FrontendTemplate.php
Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/FrontendTemplate.php:82]: fixed in 2dbedf0

**Comment by leofeyer** [comment]
@aschempp If you rebase this PR onto the latest master branch, Codecov will stop posting comments. 😉

File: core-bundle/src/EventListener/SubrequestCacheListener.php
Comment (aschempp) [reply core-bundle/src/EventListener/SubrequestCacheListener.php]: changed in e50d1d864674632ced16783d6731d4f37832cf00

File: core-bundle/src/EventListener/SubrequestCacheListener.php
Comment (aschempp) [reply core-bundle/src/EventListener/SubrequestCacheListener.php]: added in e50d1d864674632ced16783d6731d4f37832cf00

File: core-bundle/src/EventListener/SubrequestCacheListener.php
Comment (aschempp) [reply core-bundle/src/EventListener/SubrequestCacheListener.php]: changed in e50d1d864674632ced16783d6731d4f37832cf00

File: core-bundle/src/EventListener/SubrequestCacheListener.php
Comment (aschempp) [reply core-bundle/src/EventListener/SubrequestCacheListener.php:46]: theoretically, we should not need a stack, as every master request should stack and pop the strategy. But I'm not sure if there could a reason this is bypassed (e.g. an exception?), so a stack will make sure even a stacking of main requests will work.

Comment (Toflar) [reply core-bundle/src/EventListener/SubrequestCacheListener.php:46]: Okay, but we should still reset the stack, no?

Comment (aschempp) [reply core-bundle/src/EventListener/SubrequestCacheListener.php:46]: we do? https://github.com/contao/contao/pull/718/files#diff-c53e2a2fc44f16c19b46cb6d977e7ccfR73

Comment (Toflar) [reply core-bundle/src/EventListener/SubrequestCacheListener.php:46]: I meant service resetting :)

Comment (aschempp) [reply core-bundle/src/EventListener/SubrequestCacheListener.php:46]: implemented in c0ce1fd653459b850b7553e893061372870dc093

**Review by Toflar — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

File: core-bundle/src/Resources/contao/classes/FrontendTemplate.php
Comment (xchs) [inline core-bundle/src/Resources/contao/classes/FrontendTemplate.php]: Is this a typo or was it meant like that ("unsued")? 🙈 

```suggestion
	 * @param bool $blnCheckRequest 	  If true, check for unused $_GET parameters
```

Comment (leofeyer) [reply core-bundle/src/Resources/contao/classes/FrontendTemplate.php]: See #1114

**Pull request merged**
